Fix 968 display context update from

josephw opened the pull request:
Add some class checking in `update_from` to avoid an `AttributeError` when the other context includes a `_FixedPrecisionContext`.

josephw committed:
Add failing test.

josephw commented:
This change gets `linting / mypy` passing. `linting / lint` fails due to new `ruff` checks since the last green build; see #969 for fixes.

blais commented:
Hmm, I already made a fix. Please update to the latest and keep your test. I like some of the bits in here over what I did.
Please merge

josephw committed:
Deal with a _FixedPrecisionContext in update_from.

Fixes #968

josephw committed:
Remove potentially unsafe casts.

Remove just the casting from df4b77fe0ff7dd2; the new test demonstrates
that it doesn't always hold.

josephw committed:
make lint.

josephw committed:
Include necessary types for beancount/scripts/example.py.

josephw commented:
I've rebased; that involved reverting the `cast`, but otherwise it's on top of what's in `master`.

Also a couple of build fixes, to get everything [except the 3.9 builds](https://github.com/beancount/beancount/pull/967#issuecomment-3153035074) green.

josephw commented:
Rebased again, on top of the 3.9 fixes in `master`, meaning all builds are green; ready to merge.
